Migrate LogBlockIndex to RocksDb

MiguelPires opened the pull request:
A couple of relevant implementation choices about this PR:
- The previous block index worked by doing a binary search of array of entries containing pairs of log positions and "physical" addresses and returning the first entry that was smaller or equal. This means that given an event's position, the returned value should be the address of the block that the event belongs to. To reproduce the same behavior, the index iterates through the sorted positions until it reaches a block the position couldn't belong to and returns the previous one. This isn't ideal ideal but I didn't see a way of implementing a binary search with the RocksDB api and my local tests didn't show any performance problems with over a million instances created.
- The LogBlockIndex's methods are synchronized to prevent problems due to concurrent access in the RocksDb wrapper classes. I spent some time on this problem but any other solutions are probably out of the scope of this PR since they require more discussion and code. I created an issue for this #2094 
- Some changes to the data folders were made but since the docs were already outdated, I created an issue to update them #2079

closes #1730

camunda-jenkins commented:
This pull request **introduces 2 alerts** and **fixes 1** when merging edd0b3430679a1375747d032466b3406bbb29ecf into 754a6290d63145a06c6891410c5515a3f76566f0 - [view on LGTM.com](https://lgtm.com/projects/g/zeebe-io/zeebe/rev/pr-a8e23a70fa7f53bb598761d1847ef8e67aea0b74)

**new alerts:**

* 1 for Unused format argument
* 1 for Missing format argument

**fixed alerts:**

* 1 for Potential output resource leak

---

*Comment posted by [LGTM.com](https://lgtm.com)*

camunda-jenkins commented:
This pull request **fixes 1 alert** when merging 2fea17dc4649ead1714008141439ce3a13bb3ce9 into 754a6290d63145a06c6891410c5515a3f76566f0 - [view on LGTM.com](https://lgtm.com/projects/g/zeebe-io/zeebe/rev/pr-4ffbad3d5e7bb582358405773e7d00113806962c)

**fixed alerts:**

* 1 for Potential output resource leak

---

*Comment posted by [LGTM.com](https://lgtm.com)*

Zelldon commented on logstreams/src/main/java/io/zeebe/logstreams/impl/LogBlockIndexWriter.java:
```diff
@@ -136,38 +139,30 @@ protected void onActorStarting() {
       LOG.error("Failed to recover block index", e);
       throw new RuntimeException("Failed to recover block index", e);
     }
+    blockIndex.openDb();
```
why you moved this outside the try catch ? :sweat_smile:

MiguelPires replied:
Yeah, that's wrong in more than one way =p I rewrote it to catch an exception in the snapshot recovery, attempt to rebuild and then fail if an exception occurs during the rebuild

Zelldon commented on logstreams/src/main/java/io/zeebe/logstreams/impl/log/index/LogBlockIndex.java:
```diff
@@ -42,222 +41,163 @@
+    }
 
-    newBuffer.putBytes(0, indexBuffer.byteArray());
+    final File snapshotDir = getLatestSnapshotDirectory(name, snapshotConfiguration);
```
Seems really similar of the StateSnapshotController, maybe we can somehow reuse some code from there? Furthermore maybe we can use the fact that the snapshots contain the  snapshot position to find the latest snapshot.
This is also done in the StateSnapshotController:

```java

          snapshots.stream()
              .sorted(Comparator.reverseOrder())
              .filter(filter)
              .findFirst()
              .orElse(null);
```

Zelldon commented on logstreams/src/main/java/io/zeebe/logstreams/impl/log/index/LogBlockIndex.java:
```diff
@@ -42,222 +41,163 @@
-    indexBuffer.putLong(entryLogPositionOffset(entryOffset), logPosition);
-    indexBuffer.putLong(entryAddressOffset(entryOffset), storageAddr);
+    lastVirtualPosition = blockPosition;
+    key.wrapLong(blockPosition);
```
maybe we rename the key value fields to blockPosition and blockAddress?

MiguelPires replied:
Sure. I had it like it initially but thought it could be confusing having the parameters and fields be named the same way.

Zelldon commented on logstreams/src/test/java/io/zeebe/logstreams/log/LogBlockIndexWriterTest.java:
```diff
@@ -282,12 +305,5 @@ private long getSnapshotCount() {
-      return null;
-    }
-  }
+  private void getLatestSnapshot() {}
```
please remove this method

Zelldon commented on util/src/main/java/io/zeebe/util/FileUtil.java:
```diff
@@ -139,4 +141,59 @@ public static void deleteFile(File file) {
     }
   }
+
+  public static void copySnapshot(File runtimeDirectory, File snapshotDirectory) throws Exception {
```
:+1:

Zelldon commented on logstreams/src/test/java/io/zeebe/logstreams/log/LogBlockIndexTest.java:
```diff
@@ -15,85 +15,82 @@
-      assertThat(blockIndex.getAddress(i)).isEqualTo(addrFunc.apply(i));
-    }
+    lookupAndAssert(numBlocks);
+    assertThat(lastPosition).isEqualTo(lastPosition);
```
this will always be equal right? :sweat_smile:

Zelldon replied:
please assert that blockIndex.isEmpty == false

MiguelPires replied:
:man_facepalming: indeed, the first should be getLastPosition()

MiguelPires commented on broker-core/src/main/java/io/zeebe/broker/clustering/base/partitions/PartitionInstallService.java:
```diff
@@ -100,14 +101,20 @@ public void start(final ServiceStartContext startContext) {
             .logName(logName)
-            .snapshotStorage(LogStreams.createFsSnapshotStore(snapshotPath).build())
+            .blockIndexDirectory(configuration.getBlockIndexDirectory())
+            .snapshotDirectory(configuration.getSnapshotsDirectory())
```
You're right, some of those are unnecessary. I removed all of those in the latest iteration and replaced them with stateStorage which is only used by LogBlockIndex

MiguelPires commented:
Hey @Zelldon, thanks for the feedback. When you have time, please have a look at the latest iteration. I replaced the snapshot/recovery by the StateSnapshotController with some slight modifications and removed the previous code as well as FsSnapshotStorageConfiguration

camunda-jenkins commented:
This pull request **fixes 1 alert** when merging 49c76ac14dd246255605d01d677430ac534a675d into 8a033b2a5d68bf766632907af81ffd133cbaec28 - [view on LGTM.com](https://lgtm.com/projects/g/zeebe-io/zeebe/rev/pr-0aed5fd6b59b9ca6a83ae60023b6db41dc28d905)

**fixed alerts:**

* 1 for Potential output resource leak

---

*Comment posted by [LGTM.com](https://lgtm.com)*

MiguelPires commented:
bors r+

MiguelPires commented:
bors retry

Zelldon commented on broker-core/src/test/java/io/zeebe/broker/util/TestStreams.java:
```diff
@@ -107,7 +109,7 @@
   protected StateStorageFactory stateStorageFactory;
 
   public TestStreams(
-      final File storageDirectory,
```
This causes a compile error on staging, please update your branch

MiguelPires commented:
bors r+

MiguelPires commented:
bors r+

camunda-jenkins commented:
This pull request **fixes 1 alert** when merging 28af913bd42a0a4a6915a422b72fc138eaa2900b into c6bd8bc5ffc6d9f014f0d58419b525666e18056e - [view on LGTM.com](https://lgtm.com/projects/g/zeebe-io/zeebe/rev/pr-fedacfef8da975a22be6ecf77b9ae246ead5c1ac)

**fixed alerts:**

* 1 for Potential output resource leak

---

*Comment posted by [LGTM.com](https://lgtm.com)*
